client/src/components/logViewer/logParser.ts

```diff
@@ -0,0 +1,438 @@
+        ]);
+
+        return;
+      }
+
+      if (
+        isProblemIndexIndicatorLogLine(logItem.line, problemMetadata.location)
+      ) {
+        const problemIndex = getProblemIndexFromIndicatorLogLine(logItem.line);
+
+        if (problemIndex !== null) {
+          updateProblemMetadata(problemMetadata, [
+            ["problemIndicsForSameLocation", problemIndex, "amend"],
+          ]);
+        }
+
+        return;
+      }
+
+      updateProblemMetadata(problemMetadata, [
+        ["logLines", logItem.line, "amend"],
+      ]);
+
+      return;
+    } else {
+      handleProblemMetadata(
+        problems,
+        problemMetadata,
+        currentCodeLogLine,
+        offset,
+      );
+
+      if (isSourceTypeLog(logItem) && !isEmptyCodeLogLine(logItem.line)) {
+        currentCodeLogLine = logItem.line;
+
+        if (!isValidOffset(offset)) {
+          offset = calculateLocationOffset(logItem.line, firstCodeLine);
+        }
+
+        return;
+      }
+    }
+  });
+
+  handleProblemMetadata(problems, problemMetadata, currentCodeLogLine, offset);
+
+  return problems;
+}
+
+function filterOutTheLatestLogs(
+  logs: LogItem[],
+  firstCodeLine: string,
+): LogItem[] {
+  let beginningIndex = -1;
+  logs.forEach((logItem, index) => {
+    if (logItem.type !== "source") {
+      return;
+    }
+
+    const code = extractCodeInfoFromCodeLogLine(logItem.line)?.code ?? null;
+    if (code !== null && firstCodeLine === code.trim()) {
+      beginningIndex = index;
+    }
+  });
+
+  return beginningIndex === -1 ? [] : logs.slice(beginningIndex);
+}
+
+function filterOutProblemRelatedLogs(logs: LogItem[]): LogItem[] {
+  return logs.filter((logItem) => {
+    return (
+      logItem.type === "error" ||
+      logItem.type === "warning" ||
+      logItem.type === "source"
+    );
+  });
+}
+
+function updateProblemMetadata(
+  problemMetadata: ProblemMetadata,
+  data: [
+    key: string,
+    value: string | string[] | ProblemType | ProblemLocation,
+    operation: "amend" | "replace",
+  ][],
+) {
+  data.forEach((item) => {
+    const [key, value, operation] = item;
+    if (operation === "replace") {
+      problemMetadata[key] = value;
+    } else {
+      problemMetadata[key].push(value);
+    }
+  });
+}
+
+function resetProblemMetadata(problemMetadata: ProblemMetadata) {
+  removeHandledProblemIndex(problemMetadata);
+
+  const isClear = problemMetadata.problemIndicsForSameLocation.length === 0;
+  const { codeLogLine, location } = isClear
+    ? {
+        location: { startColumn: -1, endColumn: -1, lineNumber: -1 },
+        codeLogLine: "",
+      }
+    : problemMetadata;
+
+  updateProblemMetadata(problemMetadata, [
+    ["location", location, "replace"],
+    ["codeLogLine", codeLogLine, "replace"],
+    ["logLines", [], "replace"],
+    ["type", "error", "replace"],
+  ]);
+}
+
+function removeHandledProblemIndex(problemMetadata: ProblemMetadata) {
+  if (problemMetadata.problemIndicsForSameLocation.length > 0) {
```

Comment:
Can we bail early like the below, and unindent the rest
```suggestion
  if (problemMetadata.problemIndicsForSameLocation.length === 0) {
    return;
  }
```

Reply:
Done